test/array_lifetime.cpp

```diff
@@ -176,11 +196,16 @@ void test_move_lifetime(bool alloc_movable = true) {
 }
 
 TEST(array_move_lifetime) {
-  test_move_lifetime<std_alloc>();
-  test_move_lifetime<custom_alloc>();
-  test_move_lifetime<stateful_movable_custom_alloc>();
-  test_move_lifetime<auto_alloc_big>(false);
-  test_move_lifetime<auto_alloc_small>();
+  movable_custom_alloc::reset_lifetime_counters();
+  immovable_custom_alloc::reset_lifetime_counters();
+  test_move_lifetime(std_alloc());
+  test_move_lifetime(movable_custom_alloc());
+  test_move_lifetime(immovable_custom_alloc());
+  test_move_lifetime(movable_custom_alloc(1));
+  test_move_lifetime(auto_alloc_big(), false);
+  test_move_lifetime(auto_alloc_small());
+  ASSERT_EQ(movable_custom_alloc::constructs, 8);  // TODO: This seems high. The number below is correct (4 in the test itself, one in the caller).
```

Comment:
8 calls looks good to me. You have two calls where a movable custom alloc is created in the test itself:

test_move_lifetime(movable_custom_alloc());
test_move_lifetime(movable_custom_alloc(1));

and test_move_lifetime constructs a moveable allocator 3 times:

- lifetime_array<Alloc> source(lifetime_shape, alloc); --> copy constructor of Alloc through `explicit array(const Shape& shape, const Alloc& alloc)`
- lifetime_array<Alloc> source(lifetime_shape); (create a new array + alloc)
- lifetime_array<Alloc> dest(lifetime_subshape); (another new array + alloc)

The important line lifetime_array<Alloc> move(std::move(source)); doesn't create a new allocator so this looks good to me, but maybe worth splitting up into a separate test...

Reply:
Oh right, I missed movable_custom_alloc(1) and its associated tests.